### ISSUE TITLE
Hexagon generation seems to not working correctly

### ISSUE THREAD

**Issue by JohnTheCoolingFan** [StarArawn/bevy_ecs_tilemap#414]
I'm using `generate_hexagon` to select tiles that are in range for player to see. But this is what I get:
![image](https://user-images.githubusercontent.com/43478602/230120286-56960ca1-f66e-4379-b38f-981592374f1e.png)

It's supposed to be just a hexagon, but every other row below the center is shifted!

Here are code snippets of the project:
```rs
fn chart_map(
    player: Query<(&MapPos, &ChartRange), With<PlayerVehicle>>,
    mut tiles: Query<(&mut TileVisibility, &TilePos, &TilemapId)>,
    chunks: Query<&Chunk>,
) {
    let (player_pos, chart_range) = player.single();
    let tiles_in_chart_range: Vec<RowEvenPos> =
        generate_hexagon(player_pos.pos.into(), chart_range.0)
            .into_iter()
            .map(Into::into)
            .collect();
    for (mut tile_vis, tile_pos, tilemap_id) in tiles.iter_mut() {
        let chunk = chunks.get(tilemap_id.0).unwrap();
        let global_tile_pos = global_from_chunk_and_local(chunk.pos, *tile_pos);
        if tiles_in_chart_range.contains(&global_tile_pos) {
            *tile_vis = TileVisibility::Visible
        } else if matches!(*tile_vis, TileVisibility::Visible) {
            *tile_vis = TileVisibility::Charted
        }
    }
}

pub fn global_from_chunk_and_local(chunk: IVec2, local: TilePos) -> RowEvenPos {
    RowEvenPos {
        q: chunk.x * TILEMAP_CHUNK_SIZE.x as i32 + local.x as i32,
        r: chunk.y * TILEMAP_CHUNK_SIZE.y as i32 + local.y as i32,
    }
}
```

And also the project is on github: https://github.com/JohnTheCoolingFan/sands_of_merkhyl/tree/b69881fde365a79716ed8cb02999443350f8a832

**Comment by JohnTheCoolingFan** [comment]
Then this makes me think, maybe there's also problems with other AxialPos related conversions? This would probably need further testing, currently thinking on how to set it up.

**Comment by rparrett** [comment]
Yeah, I think they are all suspect. Especially anything using `ceiled_division_by_2`, which seems obviously wrong to me.

I think maybe we aren't seeing this in the `hex_neighbors` example because there are aren't any negative `q` coordinates there?

It would be great if we could rework that example so that we can repro there.

**Comment by JohnTheCoolingFan** [comment]
`hex_neighbors` example doesn't use this at all, it just `HexNeighbors` struct, which doesn't output `AxialPos`, but `TilePos` instead and can't handle tiles that are not on the map. So maybe it requires additional example?

**Comment by rparrett** [comment]
Just adding some actual rust tests for those conversion functions would probably be better.

I'm personally more inclined to remove/simplify existing hex examples actually.

**Comment by JohnTheCoolingFan** [comment]
IMO some of these examples show some important details of hex tilemap. Besides, I think it's easier to verify the system visually than figuring out valid coordinates, especially with hex. Although it is more robust and cheaper in developer time.
